File: Graph/BFS/bfs.cpp
#include<iostream>
#include<queue>

// BFS traversal

void print(bool** edges, int n) {
    bool* visited {new bool[n]()};
    int sv{0};
    std::queue<int> pendingVertex;
    pendingVertex.push(sv);
    visited[sv] = true;
    while(!pendingVertex.empty()){
        int currentVertex = pendingVertex.front();
        pendingVertex.pop();
        std::cout << currentVertex << " ";
        for(int i = 0; i<n; i++){
            if(i == currentVertex){
                continue;
            }
            if(edges[currentVertex][i] == 1 && !visited[i]){
                pendingVertex.push(i);
                visited[i] = true;
            }
        }
    }
    
    delete visited;
}


int main(){
    int n, e;
    std::cin >> n >> e;

    bool** edges {new bool*[n]};
    for (int i{}; i<n; i++) {
        edges[i] = new bool[n];
        for (int j{}; j<n; j++) {
            edges[i][j] = false;
        }
    }

    for (int i{}; i<e; i++) {
        int sv, ev;
        std::cin >> sv >> ev;
        edges[sv][ev] = true;
        edges[ev][sv] = true;
    }

    print(edges,n);

    for (int i{}; i<n; i++) {
        delete [] edges[i];
    }

    delete [] edges;
}

File: Graph/BFS/bfs.js
(function(){

    class Queue {
        #arr = [];

        push(data) {
            this.#arr.push(data);
        }

        pop(){
            this.#arr.shift();
        }

        front(){
            return this.#arr[0];
        }

        empty() {
            return this.#arr.length === 0;
        }

        print(){
            console.log(this.#arr);
        }
    }


    function printBFS(edges, n) {
        const visited = [];
        for (let i = 0; i<n; i++) visited[i] = false;

        let sv = 0;
        const pendingVertices = new Queue();
        pendingVertices.push(sv);
        visited[sv] = true;

        while(!pendingVertices.empty()) {
            let currentVertex = pendingVertices.front();
            pendingVertices.pop();
            console.log(currentVertex);

            for (let i = 0; i<n; i++) {
                if (i === currentVertex) continue;

                if (edges[currentVertex][i] && !visited[i]) {
                    visited[i] = true;
                    pendingVertices.push(i);
                }
            }
        }
    }


    (function main () {

        const n = 4;
        const e = 4;

        const connections = [
            {sv: 0, ev: 1},
            {sv: 0, ev: 3},
            {sv: 1, ev: 2},
            {sv: 2, ev: 3},
        ];


        const edges = [];
        for (let i = 0; i<n; i++) {
            edges[i] = [];
            for (let j = 0; j<n; j++) {
                edges[i][j] = false;
            }
        }

        for (let i = 0; i<e; i++){
            let {sv, ev} = connections[i];
            edges[sv][ev] = true;
            edges[ev][sv] = true;
        }

        printBFS(edges,n);

    }());


}());
